src/tools/roc_recv/main.cpp

```diff
@@ -234,19 +208,63 @@ int main(int argc, char** argv) {
     receiver_config.common.enable_profiling = args.profiling_flag;
     receiver_config.common.enable_beeping = args.beep_flag;
 
-    if (args.io_latency_given) {
-        if (!core::parse_duration(args.io_latency_arg, io_config.latency)) {
-            roc_log(LogError, "invalid --io-latency");
+    node::ContextConfig context_config;
+
+    if (args.max_packet_size_given) {
+        if (!core::parse_size(args.max_packet_size_arg, context_config.max_packet_size)) {
+            roc_log(LogError, "invalid --max-packet-size");
+            return 1;
+        }
+        if (context_config.max_packet_size == 0) {
+            roc_log(LogError, "invalid --max-packet-size: should be > 0");
             return 1;
         }
     }
 
-    if (args.rate_given) {
-        if (args.rate_arg <= 0) {
-            roc_log(LogError, "invalid --rate: should be > 0");
+    if (args.max_frame_size_given) {
+        if (!core::parse_size(args.max_frame_size_arg, context_config.max_frame_size)) {
+            roc_log(LogError, "invalid --max-frame-size");
             return 1;
         }
-        io_config.sample_spec.set_sample_rate((size_t)args.rate_arg);
+        if (context_config.max_frame_size == 0) {
+            roc_log(LogError, "invalid --max-frame-size: should be > 0");
+            return 1;
+        }
+    } else {
+        size_t rate = 48000;
+        if (args.rate_given) {
+            rate = io_config.sample_spec.sample_rate();
+        }
+        core::nanoseconds_t frame_length = sndio::DefaultFrameLength;
+        if (args.frame_len_given) {
+            frame_length = io_config.frame_length;
+        }
+        // TODO(gh-86): use channel count from --io-encoding if it's provided
+        context_config.max_frame_size =
+            (size_t)(2 * (int64_t)rate * frame_length / core::Second);
```

Comment:
Here's a bit cleaner version of this piece, what do you think?

```cpp
        audio::SampleSpec spec = io_config.sample_spec;
        if (spec.sample_rate() == 0) {
            spec.set_sample_rate(48000);
        }
        if (spec.num_channels() == 0) {
            spec.set_channel_set(audio::ChannelSet(audio::ChanLayout_Surround,
                                                   audio::ChanOrder_Smpte,
                                                   audio::ChanMask_Surround_7_1_4));
        }
        context_config.max_frame_size =
            spec.ns_2_samples_overall(io_config.frame_length) * sizeof(audio::sample_t);
```

Here we get io_config.sample_spec (filled above) and populate missing fields with values which are very rarely exceeded (48000Hz and 12 channels).

Reply:
Must admit, I was leaning a bit on you to help me out here. I'll use that code.

Reply:
So, right, what we are saying here is that... if the user hasn't supplied `--max-frame-size` then we want to calculate one that supports the maximum one.

Do we need a todo for the sample size?

Reply:
OK. I've tested with this change:
```
nolan@nolan-veed:~/nolan-veed/roc-toolkit/bin/x86_64-pc-linux-gnu$ roc-recv -vv -s rtp+rs8m://localhost:10001 -r rs8m://localhost:10002 -c rtcp://localhost:10003
13:39:29.406 [571677] [dbg] roc_core: [pool_impl.cpp:53] pool: initializing: name=packet_pool object_size=680 min_slab=8B(1S) max_slab=0B(0S)
13:39:29.406 [571677] [dbg] roc_core: [pool_impl.cpp:53] pool: initializing: name=buffer_pool object_size=2072 min_slab=8B(1S) max_slab=0B(0S)
13:39:29.406 [571677] [dbg] roc_core: [pool_impl.cpp:53] pool: initializing: name=buffer_pool object_size=4120 min_slab=8B(1S) max_slab=0B(0S)
13:39:29.406 [571677] [dbg] roc_core: [pool_impl.cpp:53] pool: initializing: name=format_pool object_size=256 min_slab=4096B(16S) max_slab=0B(0S)
13:39:29.407 [571677] [dbg] roc_node: [context.cpp:24] context: initializing
13:39:29.407 [571677] [dbg] roc_sndio: [pulseaudio_backend.cpp:21] pulseaudio backend: initializing
13:39:29.407 [571677] [dbg] roc_sndio: [sox_backend.cpp:164] sox backend: initializing
13:39:29.407 [571678] [dbg] roc_netio: [network_loop.cpp:278] network loop: starting event loop
13:39:29.407 [571679] [dbg] roc_ctl: [control_task_queue.cpp:95] control task queue: starting event loop
13:39:29.408 [571677] [dbg] roc_sndio: [backend_map.cpp:20] backend map: initialized: n_backends=2 n_drivers=82
13:39:29.408 [571677] [dbg] roc_sndio: [pulseaudio_device.cpp:75] pulseaudio sink: opening device: device=default
13:39:29.409 [571680] [inf] roc_sndio: [pulseaudio_device.cpp:507] pulseaudio sink: opening stream: device=(null) n_channels=2 sample_rate=48000
13:39:29.410 [571677] [dbg] roc_core: [pool_impl.cpp:53] pool: initializing: name=slot_pool object_size=608 min_slab=8B(1S) max_slab=0B(0S)
13:39:29.410 [571677] [dbg] roc_node: [receiver.cpp:34] receiver node: initializing
13:39:29.410 [571677] [dbg] roc_node: [receiver.cpp:75] receiver node: configuring audiosrc interface of slot 0
13:39:29.410 [571677] [inf] roc_pipeline: [receiver_source.cpp:64] receiver source: adding slot
13:39:29.410 [571677] [dbg] roc_pipeline: [receiver_slot.cpp:35] receiver slot: initializing
13:39:29.410 [571677] [inf] roc_node: [receiver.cpp:123] receiver node: binding audiosrc interface of slot 0 to rtp+rs8m://localhost:10001
13:39:29.410 [571677] [dbg] roc_pipeline: [receiver_slot.cpp:41] receiver slot: adding audiosrc endpoint rtp+rs8m
13:39:29.410 [571678] [dbg] roc_netio: [udp_receiver_port.cpp:115] udp receiver: <udprecv 0x7f3ea8001158 bind=127.0.0.1:10001>: opened port
13:39:29.410 [571677] [dbg] roc_node: [receiver.cpp:75] receiver node: configuring audiorpr interface of slot 0
13:39:29.410 [571677] [inf] roc_node: [receiver.cpp:123] receiver node: binding audiorpr interface of slot 0 to rs8m://localhost:10002
13:39:29.411 [571677] [dbg] roc_pipeline: [receiver_slot.cpp:41] receiver slot: adding audiorpr endpoint rs8m
13:39:29.411 [571678] [dbg] roc_netio: [udp_receiver_port.cpp:115] udp receiver: <udprecv 0x7f3ea80014c8 bind=127.0.0.1:10002>: opened port
13:39:29.411 [571677] [dbg] roc_node: [receiver.cpp:75] receiver node: configuring audioctl interface of slot 0
13:39:29.411 [571677] [inf] roc_node: [receiver.cpp:123] receiver node: binding audioctl interface of slot 0 to rtcp://localhost:10003
13:39:29.411 [571677] [dbg] roc_pipeline: [receiver_slot.cpp:41] receiver slot: adding audioctl endpoint rtcp
13:39:29.411 [571678] [dbg] roc_netio: [udp_receiver_port.cpp:115] udp receiver: <udprecv 0x7f3ea8001838 bind=127.0.0.1:10003>: opened port
13:39:29.411 [571677] [dbg] roc_sndio: [pump.cpp:55] pump: starting main loop
13:39:29.411 [571677] [dbg] roc_pipeline: [pipeline_loop.cpp:467] pipeline loop: tasks=4 in_place=1.00 in_frame=0.00 preempts=0 sched=0/0
13:39:29.411 [571677] [dbg] roc_sndio: [pulseaudio_device.cpp:802] pulseaudio sink: io_latency=0(0.000ms)
13:39:35.689 [571677] [inf] roc_pipeline: [receiver_session_group.cpp:217] session group: creating session: src_addr=127.0.0.1:43211 dst_addr=127.0.0.1:10001
13:39:35.689 [571677] [dbg] roc_packet: [delayed_reader.cpp:23] delayed reader: initializing: delay=8820
13:39:35.689 [571677] [dbg] roc_fec: [openfec_decoder.cpp:41] openfec decoder: initializing: codec=rs m=8
13:39:35.689 [571677] [dbg] roc_audio: [depacketizer.cpp:51] depacketizer: initializing: n_channels=2
13:39:35.689 [571677] [dbg] roc_audio: [watchdog.cpp:71] watchdog: initializing: max_blank_duration=88200 max_drops_duration=88200 drop_detection_window=13230
13:39:35.689 [571677] [dbg] roc_audio: [resampler_speex.cpp:85] speex resampler: initializing: quality=5 frame_size=32 channels_num=2
13:39:35.689 [571677] [dbg] roc_audio: [latency_monitor.cpp:73] latency monitor: initializing: target_latency=8820(200.000ms) in_rate=44100 out_rate=48000 fe_enable=1 fe_profile=gradual fe_interval=5.011ms
13:39:35.689 [571677] [dbg] roc_audio: [freq_estimator.cpp:75] freq estimator: initializing: P=1.000000e-06 I=5.000000e-09 dc1=10 dc2=10
13:39:35.690 [571677] [dbg] roc_packet: [router.cpp:63] router: detected new stream: source=1212977499 flags=0x10
13:39:35.690 [571677] [dbg] roc_rtcp: [session.cpp:42] rtcp session: initialized: is_sender=0 is_receiver=1 ssrc=2027956539 cname=TODO
13:39:35.690 [571677] [dbg] roc_rtp: [timestamp_injector.cpp:70] timestamp injector: received mapping: old=unix:0/rtp:0 new=unix:1700746775679595458/rtp:1400924589 has_ts=0 n_drops=0
13:39:35.690 [571677] [dbg] roc_audio: [resampler_speex.cpp:269] speex resampler: ratio_num=147 ratio_den=160 in_rate=44100 out_rate=48000 in_latency=40 out_latency=44
13:39:35.690 [571677] [dbg] roc_audio: [depacketizer.cpp:309] depacketizer: ts=16 loss_ratio=0.00000
13:39:35.690 [571677] [dbg] roc_audio: [watchdog.cpp:241] watchdog: status: bbbbbbbbbbbbbbbbbbbb

...

13:39:37.676 [571677] [dbg] roc_audio: [watchdog.cpp:241] watchdog: status: bbbbbbbbbbbbbbbbbbbb
13:39:37.687 [571677] [dbg] roc_audio: [watchdog.cpp:146] watchdog: blank timeout reached: every frame was blank during timeout: curr_read_pos=88208 last_pos_before_blank=0 max_blank_duration=88200
13:39:37.687 [571677] [dbg] roc_audio: [watchdog.cpp:241] watchdog: status: bbbbbbbbbbbbb
13:39:37.687 [571677] [inf] roc_pipeline: [receiver_session_group.cpp:241] session group: removing session
13:39:39.417 [571677] [dbg] roc_sndio: [pulseaudio_device.cpp:802] pulseaudio sink: io_latency=35666668(35.667ms)
13:39:49.417 [571677] [dbg] roc_sndio: [pulseaudio_device.cpp:802] pulseaudio sink: io_latency=35708336(35.708ms)
13:39:59.415 [571677] [dbg] roc_sndio: [pulseaudio_device.cpp:802] pulseaudio sink: io_latency=35854168(35.854ms)

```

Can we also add a python test to CI, that will simply spawn the applications and parse any outputs?

Reply:
> So, right, what we are saying here is that... if the user hasn't supplied --max-frame-size then we want to calculate one that supports the maximum one.

Yes, thought it's not absolute maximum, it's value that covers most of the cases. Typically rates are 44100 and 48000 and channels are 1, 2, or up to 12 in surround mode.

However rate can be also 96000 and we support up to 1024 channels. This cases are rare, and also most likely in this cases user will explicitly specify rate and channels via command line. However it's still possible that our computed maximum will be insufficient in some cases.

> Do we need a todo for the sample size?

Nope. This size is for frame pool, which is used for raw frames (audio::Frame). Unlike packets, they always use native format (float32).

Reply:
> Can we also add a python test to CI, that will simply spawn the applications and parse any outputs?

Oh, yes, this is what we're missing.

We have a task for this: https://github.com/roc-streaming/roc-toolkit/issues/552

And even an (unfinished) PR: https://github.com/roc-streaming/rt-tests/pull/8

The idea is to have a separate project ([rt-tests](https://github.com/roc-streaming/rt-tests)) that will cover more complicated aspects like latency, robustness, performance, working with real hw devices, etc. Currently I'm doing all these checks manually each release and it takes time (a few days usually).

Since it will cover CLI tools too, I think it's a good place for tests that you mentioned. I never had enough time for that project, so far all we have is a skeleton and a bunch of help-wanted issues :)

Reply:
> However it's still possible that our computed maximum will be insufficient in some cases.

Forgot to add: theoretically we could compute absolute maximum, e.g. 96000Hz + 1024channels. However that would increase memory consumption a lot (frames will be very big) and it's not needed in 99% of cases.